### ISSUE TITLE
Should `Eq(oo, oo)` return True?

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#19245]
Noticed this recently and saw it again in #14741. 

```
>>> Eq(x,x+1)
False
>>> Eq(oo,oo)
True
>>> oo + 1
oo
```
Either the first case should be unevaluated or else the 2nd case should evaluate to False because of the 3rd case.

**Comment by jksuom** [comment]
I think that the answer should be 'yes'. However, `Eq(x, x+1)` could be true if `x.is_finite` is true.

**Comment by smichr** [comment]
> could be false if `x.is_finite` is true

Agreed. But about equality of `oo`...since there are different forms of infinity and `oo - oo` is nan (somewhat reflective of that nature) why shouldn't `Eq(oo,oo)` remaing unevaluated or return False?

**Comment by jksuom** [comment]
There are different forms of infinity but only one `oo`. I think that `Eq(a, a)` should be true whenever `a == a` is true. `Eq` should not depend on subtraction (or other arithmetic operations) in general.

**Comment by smichr** [comment]
Then would you agree that `Eq(x, x + 1)` should only evaluate if `x.is_finite` is not None?

**Comment by jksuom** [comment]
That looks like a good idea (unless something else would turn up in tests).

**Comment by jlherren** [comment]
> I think that `Eq(a, a)` should be true whenever `a == a` is true.

Certainly in most cases yes, but not always IMHO. For example `Eq(nan, nan)` is currently False while `nan == nan` is (and must be) `True`. One could argue that `Eq(nan, nan)` should remain unevaluated or return `None`. But I don't think it should be `True`.

**Comment by smichr** [comment]
> `Eq(nan, nan)` is currently False while `nan == nan` is (and must be) `True`

This is the difference between structural identity (nan == nan) and mathematical convention. And I do lean towards doing the same with `oo` but this would make `Piecewise` conditions needing to test for infinity a little awkward: `Eq(x, oo) -> And(Ne(x, 0), Eq(1/x, 0))`

**Comment by sylee957** [comment]
I thought `nan` is not equal to `nan` was following the python convention.
So if the mathematic equality should follow the python convention, `Eq(oo, oo)` should be same as `inf == inf`

**Comment by oscarbenjamin** [comment]
The behaviour of Eq does not need to be constrained by `==`. In Python `==` (`__eq__`) is a low-level programmatic construct that affects the behaviour of things like sets and dicts and so on so it is important for that to match up with the basic expectations of a programmer. For that reason we have nan==nan in sympy. The Python float nan does not do this as it follows the IEEE conventions but I think that a lot of people now consider that a mistake.

In SymPy Eq can evaluate in whatever way we choose to be mathematically appropriate. I would rather not say that `Eq(nan, nan)` is False because that implies that it is well defined. It could be unevaluated though.

I would also prefer to leave `Eq(oo, oo)` unevaluated than `Eq(x, x+1)` (if those are deemed mutually exclusive).

**Comment by smichr** [comment]
> than `Eq(x, x+1)`

We are careful to not assume that `x` is real in `x + 1 > x`; we can be careful about the case that x might be infinite in the equality.

**Comment by asmeurer** [comment]
If Eq is meant to be a boolean I don't see why it shouldn't be True. There is only one positive infinity in my view. 

Regarding nan, we used to have nan == nan give False, but it lead to all sorts of problems due to the way SymPy uses ==. Actually `float('nan')` gives all sorts of problems in Python as well, but it occurs so rarely that few people know or worry about it. But in SymPy nan appears a lot. IMO SymPy shouldn't really have nan, but it's too late for that, but at least I don't think we should necessarily think of it as having floating point IEEE nan semantics.
